plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
+        $visitFieldsToSelect = array_merge(array('idvisit', 'location_ip'), array_keys(self::$logVisitFieldsToUpdate));
+
+        $lastId = 0;
+        do {
+            $logs = $this->dao->getVisitsWithDatesLimit($from, $to, $visitFieldsToSelect, $lastId, $segmentLimit);
+            if (!empty($logs)) {
+                $lastId = $logs[count($logs) - 1]['idvisit'];
+
+                $this->reattributeVisitLogs($output, $logs);
+            }
+        } while (count($logs) == $segmentLimit);
+    }
+
+    protected function reattributeVisitLogs(OutputInterface $output, $logRows)
+    {
+        foreach ($logRows as $row) {
+            if (!$this->isRowComplete($output, $row)) {
+                continue;
+            }
+
+            $location = $this->getVisitLocation($row);
+            $valuesToUpdate = $this->getValuesToUpdate($row, $location);
+
+            $this->onVisitProcessed($output);
+
+            $idVisit = $row['idvisit'];
+
+            if (empty($valuesToUpdate)) {
+                $this->writeIfVerbose($output, 'Nothing to update for idvisit = ' . $idVisit . '. Existing location info is same as geolocated.');
+            } else {
+                $this->writeIfVerbose($output, 'Updating visit with idvisit = ' . $idVisit . '.');
+
+                $this->dao->updateVisits($valuesToUpdate, $idVisit);
+                $this->dao->updateConversions($valuesToUpdate, $idVisit);
+            }
+        }
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return int
+     */
+    protected function getPercentStep(InputInterface $input)
+    {
+        $percentStep = $input->getArgument(self::PERCENT_STEP_ARGUMENT);
+
+        if (!is_numeric($percentStep)) {
+            return self::PERCENT_STEP_ARGUMENT_DEFAULT;
+        }
+
+        // Percent step should be between maximum percent value and minimum percent value (1-100)
+        if ($percentStep > 99 || $percentStep < 1) {
+            return 100;
+        }
+
+        return $percentStep;
+    }
+
+    /**
+     * Returns location log values that are different than the values currently in a log row.
+     *
+     * @param array $row The visit row.
+     * @param array $location The location information.
+     * @return array The location properties to update.
+     */
+    protected function getValuesToUpdate(array $row, $location)
```

Comment:
> We can also name it Gateway. In the end it shouldn't contain such code there IMO.

If I understand correctly, you don't want the command to have logic that does any real work (ie, the logic that literally carries out the command's function). My opinion is that the `execute()` + `configure()` methods are what shouldn't be there, as all they do is define the CLI interface to the logic. The result of both approaches is the same, one class w/ the business logic and another w/ the interface metadata, but yours introduces more technical debt. There will be a new class that may never be re-used or meant to be re-used, that if it is ever re-used can be easily used in an improper context or can have its logic re-implemented because it is not flexible enough, that will have tests that need to be maintained but only test what is a seldom used and simple administrative function.

In other languages, the act of tagging classes w/ metadata is achieved through attribute oriented programming, eg, in C# it might look like:

```
[Command(method = 'doSomething')]
class MyCommand
{
    public int doSomething([CommandParameter(help = "...")] String myParam,
                     [CommandParameter] String myOtherParam = "abc")
    {
        ...
    }
}
```

Since PHP + Symfony forces the metadata to be defined procedurally, and encourages the actual logic to be hidden, we run into this problem.

All this said, I understand that saying "PHP + Symfony should be better" isn't particularly useful. So, I'm going to move the logic that re-attributes a visit w/ location info into VisitorGeolocator. I won't move processSpecifiedLogsInChunks, since it literally just iterates over the log_visit table. Like I said before, this could be abstracted into an Iterator, but since all it would do is iterate over the log_visit table in chunks, it should go in core. And that would require more review, so perhaps should be done in its own pull request.

> Of course you can leave it there if you want.

I would not consider this a good way to resolve a difference of opinion. I would prefer to reach an understanding, rather than one person acting independently or capitulating unconditionally.

> Goal must be to have tests for this.

I think the current goal should be to have something usable in 2.12 for the particular client who has been waiting for a long time to have this functionality in cloud. There is a system test that tests accuracy of the command, perhaps it can be improved. Unit tests for bits and pieces can come later, since the tests you want to add do not test anything that interacts w/ Piwik's core functionality. (Eg, VisitorGeolocator is used in tracking, so it has tests now, but looping over the log_visit table and updating visits is not used anywhere but in the command, so they don't necessarily have to be tested in isolation, right at this very moment.)